File: code/algorithum/Course/linked-list/queue.js
class Node{
    constructor(value){
        this.value  = value;
        this.next = null; 
    }
}
class LinkedList{
    constructor(){
        this.head = null;
        this.tail = null;
    }
}
const linkedlist = new LinkedList();

function pushToQueue(item){
    const node = new Node(item);
    if(linkedlist.head === null) {
        linkedlist.head = node; 
        linkedlist.tail = node;
        linkedlist[node.value] = node;
    }
    else{
        linkedlist.tail.next = node;
        linkedlist[node.value] = node;  
        linkedlist.tail = node;
    }
}

function popFromQueue(){
    let temp = linkedlist.head.next;
    linkedlist.head.next = null;
    linkedlist.head = temp;
}

const array = [2,1,4,3,7];
for (let i = 0; i < array.length; i++) {
    pushToQueue(array[i]);
}


console.log(linkedlist);

popFromQueue();

console.log(linkedlist);
popFromQueue();

console.log(linkedlist);

File: code/algorithum/Course/linked-list/queue.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

class Node{
    public:
        int data;
        Node* next;
        Node(){
            data = 0;
            next = nullptr;
        }
        Node(int data){
            this->data = data;
            this->next = nullptr;
        }
};
class singlyLinkedList{
    public:
        Node* head;
        singlyLinkedList(){
            head = nullptr;
        }
        void insertAtHead(int);
        void printSinglyLinkedList(int);
};
void singlyLinkedList::insertAtHead(int data){
    Node* newNode = new Node(data);
    if(head == NULL){
        head = newNode;
        return ;
    };
    Node* temp  = head;
    newNode->next = temp;
    head = newNode;
};
void singlyLinkedList::printSinglyLinkedList(int number){
    Node* temp = head;
    for (int i = 0; i < number-1; i++)
    {
        cout << temp->data << endl;
        temp = temp->next;
    }
    
    cout << temp->data;
}

int main() {
    /* Enter your code here. Read input from STDIN. Print output to STDOUT */   
    int number,data;
    cin >> number;
    singlyLinkedList list;
    for(int i = 0; i < number; i++){
        cin >> data;
        list.insertAtHead(data);
    };

    list.printSinglyLinkedList(number);
    return 0;
}
